feat: add support for pkl files

akhil-rasheed opened the pull request:
Closes #141

Addresses following comments left on the original (abandoned) PR here https://github.com/Boeing/config-file-validator/pull/164

- Moves PKL binary validation to the pkl validator
- Ensures we don't run into race conditions during testing by using a mutex lock
- Adds a more complete example for good.pkl
- Alphabetises list of supported config filetypes

olunusib committed:
feat: add support for pkl files

olunusib committed:
feat: skip pkl files when binary isn't in PATH

olunusib committed:
Merge branch 'main' into feat/pkl-validation

olunusib committed:
chore: use the new action tag in ref comment

Commit:
fix(test): resolve data race in pkl validation test

The existing test for ignoring PKL files when the binary is missing was not safe for parallel execution, as it modified a global function variable, creating a data race condition.

This commit resolves the issue by refactoring the validation logic:

1. The responsibility for checking for the pkl binary is moved from the CLI into the `PklValidator`.
2. The validator now returns a specific `ErrSkipped` error if the binary is not found.
3. A thread-safe `SetPklBinaryChecker` function is introduced, allowing tests to safely mock the check.
4. The test is updated to use this new thread-safe mechanism.

Commit:
fix: move PKL to alphabetically correct spot in list of supported config filetypes

Commit:
test: add fleshed out example for good.pkl

akhil-rasheed committed:
Merge branch 'main' into feat/pkl-validation

Commit:
fix(test): repair test suite after pkl validator refactor

- Removes a redeclared isPklBinaryPresent function that was causing a build failure.
- Updates the test logic to check for the new ErrSkipped sentinel error, aligning it with the recent refactoring.
- Removes an unused import.

Commit:
fix: correct build and pkl test fixture

- Fixes a build failure in the filetype package by using the correct `tools` import.
- Updates the `good.pkl` test fixture with valid syntax so that tests pass when the pkl binary is installed.

kehoecj commented:
@akhil-rasheed Thanks for the PR! Please address the issues identified in the unit and mega linter jobs.

Commit:
fix: remove whitespace in workflow file

Commit:
feat(pkl): Improve test coverage for pkl validator

Refactors the pkl validator to improve testability and adds tests for uncovered code paths.

- Modifies the `PklValidator` to allow injecting an evaluator factory, making it easier to test error conditions.
- Adds a test to verify that `ErrSkipped` is returned when the `pkl` binary is not found.
- Adds a test to verify that an error is returned when the pkl evaluator fails to be created.

akhil-rasheed committed:
fix(test): Prevent race conditions in pkl validator tests

Removes `t.Parallel()` from tests that modify the global `isPklBinaryPresent`
variable. This prevents potential race conditions between tests that
rely on the state of this variable.

akhil-rasheed commented:
> > I've addressed the coverage issues and some whitespace in the workflow file which was causing the yaml linter to fail @kehoecj
> 
> Still seeing at least one test failure: [#331 (comment)](https://github.com/Boeing/config-file-validator/pull/331#issuecomment-3381095710)

Resolved

kehoecj commented:
> > > I've addressed the coverage issues and some whitespace in the workflow file which was causing the yaml linter to fail @kehoecj
> > 
> > 
> > Still seeing at least one test failure: [#331 (comment)](https://github.com/Boeing/config-file-validator/pull/331#issuecomment-3381095710)
> 
> Resolved

Can you take another look? I'm still seeing a failure in the pipeline

akhil-rasheed commented:
```
Total Summary: 35 succeeded, 0 failed
--- FAIL: Test_TripleGroupOutput (0.11s)
=== RUN   Test_IncorrectSingleGroupOutput
--- PASS: Test_IncorrectSingleGroupOutput (0.10s)
=== RUN   Test_IncorrectDoubleGroupOutput
--- PASS: Test_IncorrectDoubleGroupOutput (0.06s)
=== RUN   Test_IncorrectTripleGroupOutput
--- PASS: Test_IncorrectTripleGroupOutput (0.07s)
FAIL
coverage: 93.2% of statements
FAIL	github.com/Boeing/config-file-validator/pkg/cli	0.781s
?   	github.com/Boeing/config-file-validator/pkg/filetype	[no test files]
```

This is on the CI, I assume this is where it's failing, but the same test passes for me locally
```
Total Summary: 35 succeeded, 0 failed
--- PASS: Test_TripleGroupOutput (0.07s)
=== RUN   Test_IncorrectSingleGroupOutput
--- PASS: Test_IncorrectSingleGroupOutput (0.06s)
=== RUN   Test_IncorrectDoubleGroupOutput
--- PASS: Test_IncorrectDoubleGroupOutput (0.07s)
=== RUN   Test_IncorrectTripleGroupOutput
--- PASS: Test_IncorrectTripleGroupOutput (0.06s)
PASS
ok  	github.com/Boeing/config-file-validator/pkg/cli	1.423s
```

Am I just not identifying the failure correctly?

kehoecj commented:
> ```
> Total Summary: 35 succeeded, 0 failed
> --- FAIL: Test_TripleGroupOutput (0.11s)
> === RUN   Test_IncorrectSingleGroupOutput
> --- PASS: Test_IncorrectSingleGroupOutput (0.10s)
> === RUN   Test_IncorrectDoubleGroupOutput
> --- PASS: Test_IncorrectDoubleGroupOutput (0.06s)
> === RUN   Test_IncorrectTripleGroupOutput
> --- PASS: Test_IncorrectTripleGroupOutput (0.07s)
> FAIL
> coverage: 93.2% of statements
> FAIL	github.com/Boeing/config-file-validator/pkg/cli	0.781s
> ?   	github.com/Boeing/config-file-validator/pkg/filetype	[no test files]
> ```
> 
> This is on the CI, I assume this is where it's failing, but the same test passes for me locally
> 
> ```
> Total Summary: 35 succeeded, 0 failed
> --- PASS: Test_TripleGroupOutput (0.07s)
> === RUN   Test_IncorrectSingleGroupOutput
> --- PASS: Test_IncorrectSingleGroupOutput (0.06s)
> === RUN   Test_IncorrectDoubleGroupOutput
> --- PASS: Test_IncorrectDoubleGroupOutput (0.07s)
> === RUN   Test_IncorrectTripleGroupOutput
> --- PASS: Test_IncorrectTripleGroupOutput (0.06s)
> PASS
> ok  	github.com/Boeing/config-file-validator/pkg/cli	1.423s
> ```
> 
> Am I just not identifying the failure correctly?

That appears to be the correct failure - let me see if it passes locally for me as well. There may be a timing issue that is present in the pipeline causing it to fail

akhil-rasheed committed:
refactor: rename ErrSkipped to ErrPklSkipped

akhil-rasheed committed:
refactor: improve pkl binary detection

akhil-rasheed committed:
revert: add back validator to .gitignore

akhil-rasheed committed:
Merge branch 'main' into feat/pkl-validation

akhil-rasheed committed:
fix: remove whitespace

akhil-rasheed committed:
fix: remove whitespace

akhil-rasheed committed:
refactor(suggestion): mark unused parameters

Co-authored-by: ccoVeille <3875889+ccoVeille@users.noreply.github.com>

kehoecj reviewed (changes requested):
When the PKL executable is not installed, a runtime error is generated:

```
> .\validator.exe .\test\                         
Warning: 'pkl' binary not found, file C:\Users\se456c\src\github.com\prs\pkl\config-file-validator\test\fixtures\good.pkl will be ignored.
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x7ff78ccb90f0]

goroutine 22 [running]:
github.com/apple/pkl-go/pkl.(*execEvaluator).deinit(0xc000206140)
        C:/Users/se456c/go/pkg/mod/github.com/apple/pkl-go@v0.8.0/pkl/evaluator_manager_exec.go:214 +0x70        
github.com/apple/pkl-go/pkl.(*evaluatorManager).closeErr(0xc000193bc0, {0x7ff78cee41c0?, 0xc000116e80?})
        C:/Users/se456c/go/pkg/mod/github.com/apple/pkl-go@v0.8.0/pkl/evaluator_manager.go:308 +0x82
github.com/apple/pkl-go/pkl.(*evaluatorManager).listenForImplClose(0xc000193bc0)
        C:/Users/se456c/go/pkg/mod/github.com/apple/pkl-go@v0.8.0/pkl/evaluator_manager.go:245 +0x46
created by github.com/apple/pkl-go/pkl.NewEvaluatorManagerWithCommand in goroutine 1
        C:/Users/se456c/go/pkg/mod/github.com/apple/pkl-go@v0.8.0/pkl/evaluator_manager_exec.go:61 +0x256
```

kehoecj commented:
@akhil-rasheed Just checking in - any questions or concerns about the last set of review comments?

akhil-rasheed committed:
feat: remove pkl binary dependency

Use the in-process pkl-go interpreter for validation instead of requiring the pkl binary to be installed.

This simplifies usage by removing the external dependency.

akhil-rasheed committed:
Merge branch 'main' into feat/pkl-validation

akhil-rasheed commented on .github/workflows/go.yml:
```diff
@@ -97,6 +97,11 @@ jobs:
         with:
           go-version: "1.25"
 
+      - name: Set up pkl
```
Yep, the binary wasn't required, I have removed the dependency and related tests etc

akhil-rasheed committed:
fix: remove pkl installation in workflow

akhil-rasheed committed:
fix: remove unintented whitespace additions

akhil-rasheed committed:
Merge branch 'main' into feat/pkl-validation

akhil-rasheed committed:
feat: Implement PklValidator to satisfy Validator interface                                                                                Renamed PklValidator.Validate to PklValidator.ValidateSyntax and added PklValidator.ValidateFormat to correctly implement the Validator interface. Updated test calls in validator_test.go accordingly

ccoVeille commented:
I don't have the privilege to launch it unfortunately @akhil-rasheed 

I'm not member

I thought I had 😅

kehoecj committed:
Merge branch 'main' into feat/pkl-validation

kehoecj commented:
@akhil-rasheed Closing. This branch is so out of date with main it's going to be easier to get a clean main and retroactively apply your changes. Feel free to open a new MR with a syntax-only implementation and no binary dependency
